Allow editing workflow templates

CarinaWolli opened the pull request:
## What does this PR do?

With this PR workflow templates are made more flexible. Predefined templates can now be edited and dates can be formatted (as you can see in the screenshot). It also introduces new dynamic text variables for _cancel url_, _rescheduling_ url, _event end time_ and _timezone_. 

![Screenshot 2023-04-07 at 09 06 48](https://user-images.githubusercontent.com/30310907/230559870-1f41bfb5-17e3-429c-8e63-e6aff5c6d7a4.png)

![Screenshot 2023-04-04 at 14 14 34](https://user-images.githubusercontent.com/30310907/230311849-9635a082-ed3d-40a6-8106-1426379c1e0f.png)

Adds info message that emails and SMS can only be scheduled at least 1 hour in advance: 
![Screenshot 2023-04-03 at 16 47 55](https://user-images.githubusercontent.com/30310907/229545566-51236c47-b054-494e-ae34-827f6b9bacea.png)

Fixes custom booking questions as dynamic text variables for all triggers. They only worked when a normal event was booked. It didn’t work when an event needed confirmation or when an event was canceled
Before it was shown as `[object Object]` instead of the value

Fixes #7924 
Fixes #7496
Fixes #4944
Fixes #8001

## Type of change

- New feature (non-breaking change which adds functionality)
- Bug fix

Commit:
add event end time as variable

Commit:
Merge branch 'main' into feat/flexible-workflow-templates

Commit:
add timezone as new variable

Commit:
Merge branch 'main' into feat/flexible-workflow-templates

Commit:
add first version of template prefill

Commit:
set template body when template is updated

Commit:
Merge branch 'main' into feat/flexible-workflow-templates

Commit:
set reminder template body and subject when creating workflow

Commit:
set email subject when changes templates

Commit:
save emailBody and emailsubject for all templates + fix duplicate template text

Commit:
add more flexibility for templates

Commit:
remove console.log

Commit:
fix {ORAGANIZER} and {ATTENDEE} variable

Commit:
make sure to always send reminder body and not default template

Commit:
fix import

Commit:
remove email body text and match variables in templates

Commit:
handle translations of formatted variables

Commit:
Merge branch 'main' into feat/flexible-workflow-templates

Commit:
fix email reminder template

Commit:
add cancel and reschedule link as variable

Commit:
add cancel and reschedule link for scheduled emails/sms

Commit:
make sure empty empty body and subject are set for reminder template

Commit:
Merge branch 'main' into feat/flexible-workflow-templates

Commit:
add info message for testing workflow

Commit:
fix typo

Commit:
Merge branch 'main' into feat/flexible-workflow-templates

Commit:
add sms template

Commit:
add migration to remove reminderBody and emailSubject

Commit:
add branding

Commit:
Merge branch 'main' into feat/flexible-workflow-templates

Commit:
code clean up

Commit:
add hide branding everywhere

Commit:
Merge branch 'main' into feat/flexible-workflow-templates

Commit:
fix sms reminder template

Commit:
set sms reminder template if sms body is empty

Commit:
fix custom inputs variables everywhere

Commit:
fix variable translations + other small fixes

CarinaWolli commented on packages/ui/components/editor/plugins/ToolbarPlugin.tsx:
```diff
@@ -340,23 +340,47 @@ export default function ToolbarPlugin(props: TextEditorProps) {
-      const parser = new DOMParser();
-      const dom = parser.parseFromString(props.getText(), "text/html");
+    if (!props.firstRender) {
+      editor.update(() => {
```
fill editor with new updated text

CarinaWolli commented on packages/ui/components/editor/plugins/ToolbarPlugin.tsx:
```diff
@@ -340,23 +340,47 @@ export default function ToolbarPlugin(props: TextEditorProps) {
-      const nodes = $generateNodesFromDOM(editor, dom);
+  useEffect(() => {
+    if (props.setFirstRender) {
+      props.setFirstRender(false);
```
everything except `props.setFirstRender(false)` stayed the same in this `useEffect`

Commit:
fix some type errors

Commit:
fix more type errors

Commit:
fix everything missing around cron job scheduling

Commit:
make sure to always use custom template for sms messages

Commit:
fix type error

Commit:
code clean up

CarinaWolli commented on packages/prisma/migrations/20230404100155_remove_reminder_body_email_subject_from_reminder_template/migration.sql:
```diff
@@ -0,0 +1,6 @@
+-- Until now reminderBody and emailSubject were ignored for the REMINDER template
+-- From now on we use the reminderBody and emailSubject also for templates
+
```
Migration to make sure all reminderBodies and emailSubjects are empty for workflow steps with the REMINDER template. If reminderBody is empty we will use our normal predefined reminder template

Commit:
Merge branch 'main' into feat/flexible-workflow-templates

CarinaWolli committed:
Merge branch 'main' into feat/flexible-workflow-templates

Commit:
rename link to url

Commit:
Merge branch 'main' into feat/flexible-workflow-templates

Commit:
Merge branch 'main' into feat/flexible-workflow-templates

hariombalhara commented:
Starting reviewing this.

hariombalhara commented on packages/features/ee/workflows/lib/reminders/templates/emailReminderTemplate.ts:
```diff
@@ -1,43 +1,50 @@
+  const attendeeHtml = `<div><strong class="editor-text-bold">Attendees:</strong></div>You & ${attendee}<br><br>`;
 
-  const endingHtml = `This reminder was triggered by a Workflow in Cal.<br><br>_<br><br>Scheduling by ${APP_NAME}</body>`;
+  const branding = !isBrandingDisabled && !isEditingMode ? `<br><br>_<br><br>Scheduling by ${APP_NAME}` : "";
```
I am not sure if we should send this branding text in emails from self-hosted instances. Though it would have their APP_NAME in the text but suddenly starting sending this text might be unexpected.

hariombalhara replied:
Also, we are not showing this in the template during editing but maybe we should mention somewhere(in the app) that there would be branding text as well in all emails depending on the connected event-type's `hideBranding` status.

hariombalhara commented on packages/features/ee/workflows/lib/reminders/templates/emailReminderTemplate.ts:
```diff
@@ -1,43 +1,50 @@
+    endTime = "{EVENT_END_TIME}";
+    eventName = "{EVENT_NAME}";
+    timeZone = "{TIMEZONE}";
+    attendee = action === WorkflowActions.EMAIL_ATTENDEE ? "{ORGANIZER}" : "{ATTENDEE}";
```
nit: I think better name for this variable is something like `otherPerson`. Because it's used as `You & {otherPerson}`

hariombalhara commented on packages/features/ee/workflows/lib/reminders/templates/emailReminderTemplate.ts:
```diff
@@ -1,43 +1,50 @@
+    timeZone = "{TIMEZONE}";
+    attendee = action === WorkflowActions.EMAIL_ATTENDEE ? "{ORGANIZER}" : "{ATTENDEE}";
+    name = action === WorkflowActions.EMAIL_ATTENDEE ? "{ATTENDEE}" : "{ORGANIZER}";
+    eventDate = "{EVENT_DATE_ddd, MMM D, YYYY H:mmA}";
```
nit: I think we should abstract these variables out so that we can maintain consisteny b/w email and sms reminders.

hariombalhara commented on packages/features/ee/workflows/lib/variableTranslations.ts:
```diff
@@ -12,12 +12,29 @@ export function getTranslatedText(text: string, language: { locale: string; t: T
+        if (variable.toLowerCase().startsWith(formattedVar)) return variable;
+      })[0];
+
+      if (formattedVarToTranslate) {
```
Can we have comments on what we are trying to achieve here.

hariombalhara commented on packages/trpc/server/routers/viewer/workflows.tsx:
```diff
@@ -349,7 +350,7 @@ export const workflowsRouter = router({
             name: "",
             trigger: WorkflowTriggerEvents.BEFORE_EVENT,
-            time: 24,
+            time: 2,
```
I don't understand the reasoning for this change.

hariombalhara committed:
Add debug logs

hariombalhara committed:
Merge remote-tracking branch 'origin/feat/flexible-workflow-templates' into feat/flexible-workflow-templates

hariombalhara commented on packages/features/ee/workflows/lib/reminders/templates/customTemplate.ts:
```diff
@@ -1,54 +1,84 @@
-    .replaceAll("{EVENT_TIME}", timeWithTimeZone)
+    .replaceAll("{EVENT_TIME}", variables.eventDate?.format("H:mmA") || "")
+    .replaceAll("{EVENT_END_TIME}", variables.eventEndTime?.format("H:mmA") || "")
+    .replaceAll("{TIMEZONE_TIME}", variables.timeZone || "")
```
TIMEZONE_TIME and TIMEZONE have same values?

hariombalhara replied:
<img width="553" alt="Screenshot 2023-04-13 at 7 58 35 PM" src="https://user-images.githubusercontent.com/1780212/231791096-6188aae0-8bef-4482-b231-b1b56457c660.png">
This is confusing. 
- If an attendee receives the email it would be the timezone of attendee ?
- If an organizer receives the email it would be the timezone of organizer ?

CarinaWolli replied:
> This is confusing.
> If an attendee receives the email it would be the timezone of attendee ?
> If an organizer receives the email it would be the timezone of organizer ?

Would "The timezone of the person receiving" be better? Not quite sure how to keep it short but understandable, but I agree that this is a little confusing

hariombalhara replied:
Yeah I was thinking that only. That makes much more sense

hariombalhara commented:
<img width="970" alt="Screenshot 2023-04-13 at 7 56 54 PM" src="https://user-images.githubusercontent.com/1780212/231790503-991b7349-5ed7-405e-8233-7e5ba98d8ad5.png">
Small misalignment here

hariombalhara commented:
<img width="1123" alt="Screenshot 2023-04-13 at 8 08 23 PM" src="https://user-images.githubusercontent.com/1780212/231794648-cecfc1a1-a752-49ad-87e1-165bc2441f93.png">
Select sms action shows template like this. I believe this is unexpected.

hariombalhara reviewed (approved):
Have left some comments.
But so far I didn't find any functional bug. Awesome work @CarinaWolli !!

zomars committed:
Merge branch 'main' into feat/flexible-workflow-templates

zomars committed:
Merge branch 'main' into feat/flexible-workflow-templates

zomars committed:
Update handleNewBooking.ts

zomars reviewed (commented):
Code LGTM. Solved conflicts with @alannnc changes. Can you verify I didn't made any unintended replacements? @CarinaWolli 🙏🏽

alannnc commented on packages/features/bookings/lib/handleConfirmation.ts:
```diff
@@ -172,14 +214,21 @@ export async function handleConfirmation(args: {
         evtOfBooking.startTime = updatedBookings[index].startTime.toISOString();
         evtOfBooking.endTime = updatedBookings[index].endTime.toISOString();
         evtOfBooking.uid = updatedBookings[index].uid;
+        const eventTypeSlug = updatedBookings[index].eventType ? updatedBookings[index].eventType?.slug : "";
```
```suggestion
const eventTypeSlug = updatedBookings[index].eventType?.slug ?? "";
```

alannnc commented on packages/features/ee/workflows/lib/reminders/reminderScheduler.ts:
```diff
@@ -8,18 +8,24 @@ import type { CalendarEvent } from "@calcom/types/Calendar";
 import { scheduleEmailReminder } from "./emailReminderManager";
 import { scheduleSMSReminder } from "./smsReminderManager";
 
+type ExetendedCalendarEvent = CalendarEvent & {
```
```suggestion
type ExtendedCalendarEvent = CalendarEvent & {
```
Typo?

alannnc commented on packages/features/ee/workflows/lib/reminders/reminderScheduler.ts:
```diff
@@ -8,18 +8,24 @@ import type { CalendarEvent } from "@calcom/types/Calendar";
   })[];
   smsReminderNumber: string | null;
-  calendarEvent: CalendarEvent & { metadata?: { videoCallUrl: string | undefined } };
+  calendarEvent: ExetendedCalendarEvent;
```
```suggestion
  calendarEvent: ExtendedCalendarEvent;
```

alannnc commented on packages/features/ee/workflows/lib/reminders/reminderScheduler.ts:
```diff
@@ -111,11 +119,12 @@ export interface SendCancelledRemindersArgs {
   })[];
   smsReminderNumber: string | null;
-  evt: CalendarEvent;
+  evt: ExetendedCalendarEvent;
```
```suggestion
  evt: ExtendedCalendarEvent;
```

alannnc commented on packages/features/ee/workflows/lib/reminders/templates/customTemplate.ts:
```diff
@@ -1,54 +1,84 @@
+    .replaceAll("{TIMEZONE}", variables.timeZone || "")
+    .replaceAll("{CANCEL_URL}", cancelLink)
+    .replaceAll("{RESCHEDULE_URL}", rescheduleLink)
     .replaceAll("{MEETING_URL}", variables.meetingUrl || "");
```
I think probably we can use functions that wraps this replaceAlls.
e.g. replaceAllOldVariables()
replaceAll() // With only the new ones?

alannnc commented on packages/trpc/server/routers/viewer/workflows.tsx:
```diff
@@ -10,12 +10,14 @@ import {
 } from "@calcom/features/bookings/lib/getBookingFields";
 import type { WorkflowType } from "@calcom/features/ee/workflows/components/WorkflowListPage";
+import { isSMSAction } from "@calcom/features/ee/workflows/lib/actionHelperFunctions";
 // import dayjs from "@calcom/dayjs";
```
remove this since we are already here

alannnc committed:
Merge branch 'main' into feat/flexible-workflow-templates

alannnc reviewed (approved):
Left some nit changes, functionality works.
There was a time when I saved template field and went back and clicked item and template text was filled with my old text.

hariombalhara committed:
Add debug logs

hariombalhara committed:
Merge remote-tracking branch 'origin/feat/flexible-workflow-templates' into feat/flexible-workflow-templates

Commit:
removed unneded responses

CarinaWolli commented on packages/features/bookings/lib/handleConfirmation.ts:
```diff
@@ -155,8 +191,14 @@ export async function handleConfirmation(args: {
+        attendees: true,
+        location: true,
+        customInputs: true,
+        responses: true,
```
You are right, not needed. Removed it

Commit:
fix booking questions + UI improvements

Commit:
remove html email body when changing to sms action

Commit:
code clean up + comments

Commit:
code clean up

Commit:
code clean up

Commit:
remove comment

CarinaWolli commented:
@hariombalhara & @alannnc I addressed all your feedback, thank you for that 🙏 

Some of your feedback will be handled in a follow-up, see here: https://github.com/calcom/cal.com/issues/8257

CarinaWolli committed:
Merge branch 'main' into feat/flexible-workflow-templates

Commit:
more clear info message for timezone variable

CarinaWolli committed:
Merge branch 'main' into feat/flexible-workflow-templates

CarinaWolli commented:
@zomars does it need your approval or why can't this be merged yet? 
<img width="853" alt="Screenshot 2023-04-17 at 08 31 34" src="https://user-images.githubusercontent.com/30310907/232402815-f4714b07-8b9b-472a-bdb6-1f101a6ea60b.png">

hariombalhara committed:
Merge branch 'main' into feat/flexible-workflow-templates

hariombalhara commented:
@CarinaWolli I re-added myself as a review. It should start the tests. 
Right now it requires someone to be added as reviewer manually. @zomars mentioned it on threads

CarinaWolli commented:
Would need an approval from @calcom/ui here, maybe @sean-brydon or @PeerRich?

sean-brydon reviewed (approved):
LGTM from the UI side :) 

Email builder is really coming along well 👏
